content/library/advanced-features/advanced-widget-behavior.md

```diff
@@ -14,3 +16,252 @@ Widgets are magical and often work how you want. But they can have surprising be
+
+#### Updating a slider with a default value
+
+For the last two sliders above, a change to the min or max value will result in the widgets being seen as "new" and thus recreated like before. Since a default value of 5 is defined, each widget will reset to 5 whenever the min or max is changed. This is again the same (with or without a key).
+
+A solution to [Retain statefulness when changing a widget's parameters](#retain-statefulness-when-changing-a-widgets-parameters) is provided further on.
+
+### Widgets do not persist when not continually rendered
+
+If a widget's function is not called during a script run, then none of its parts will be retained, including its value in `st.session_state`. If a widget has a key and you navigate away from that widget, its key and associated value in `st.session_state` will be deleted. Even temporarily hiding a widget will cause it to reset when it reappears; Streamlit will treat it like a new widget. You can either interrupt the [widget cleanup process](#cleaning-up) (described at the end of this page) or save the value to another key.
+
+#### Save widget values in Session State to preserve them between pages
+
+If you want to navigate away from a widget and return to it while keeping its value, use a separate key in `st.session_state` to save the information independently from the widget. In this example, a temporary key is used with a widget. The temporary key uses an underscore prefix. Hence, `_my_key` is used as the widget key, but the data is copied to `my_key` to preserve it between pages.
+
+```python
+import streamlit as st
+
+def save_value():
+    # Copy the value to the permanent key
+    st.session_state["my_key"] = st.session_state["_my_key"]
+
+# Copy the saved value to the temporary key
+st.session_state["_my_key"] = st.session_state["my_key"]
+st.number_input("Number of filters", key="_my_key", on_change=save_value)
+```
+
+If this is functionalized to work with multiple widgets, it could look something like this:
+
+```python
+import streamlit as st
+
+def save_value(key):
+    st.session_state[key] = st.session_state["_"+key]
+def get_value(key):
+    st.session_state["_"+key] = st.session_state[key]
+
+get_value("my_key")
+st.number_input("Number of filters", key="_my_key", on_change=save_value, args="my_key")
+
+```
+
+## Widget life cycle
+
+When a widget function is called, Streamlit will check if it already has a widget with the same parameters. Streamlit will reconnect if it thinks the widget already exists. Otherwise, it will make a new one.
```

Comment:
`reconnect` feels a bit off here, makes me think of session connection. Maybe `Streamlit will reuse a matching widget if it already exists. Otherwise...` or something like that?

Reply:
Hmmm. For my edification, what would be the nitty-gritty, technically correct description of what Streamlit is doing here to match a function to existing widgets @AnOctopus?